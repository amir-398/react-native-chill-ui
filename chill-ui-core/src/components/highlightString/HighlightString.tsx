import { cn } from '@utils';
import { memo } from 'react';

import { classNamePropsHandler } from '@/utils/hybrid/classNamePropsHandler';
import { classNameHandler, styleHandler } from '@/utils/hybrid/propsHandlers';

import { String } from '../string';
import { HighlightStringProps } from '../../types/highlightString.types';
import { isNativeWindInstalled } from '../../utils/hybrid/nativewindDetector';

/**
 * HighlightString component that highlights specific terms within text.
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * // With NativeWind
 * <HighlightString
 *   text="Hello world, welcome to the world of programming"
 *   highlightTerm="world"
 *   className="text-base text-gray-800"
 *   highlightClassName="bg-yellow-200 font-bold"
 * />
 *
 * // Without NativeWind (fallback)
 * <HighlightString
 *   text="Hello world, welcome to the world of programming"
 *   highlightTerm="world"
 *   style={{ fontSize: 16, color: '#374151' }}
 *   highlightStyle={{ backgroundColor: '#FEF3C7', fontWeight: 'bold' }}
 * />
 * ```
 *
 * @param text - The full text to display
 * @param highlightTerm - The term to highlight within the text
 * @param className - Custom CSS classes for the container (NativeWind only)
 * @param highlightClassName - Custom CSS classes for highlighted text (NativeWind only)
 * @param style - Custom styles for the container (StyleSheet only)
 * @param highlightStyle - Custom styles for highlighted text (StyleSheet only)
 * @param stringProps - Props for the main string component
 * @param highlightStringProps - Props for the highlighted string component
 * @returns Text with highlighted terms
 */
function HighlightString(props: HighlightStringProps) {
  classNamePropsHandler(props, 'HighlightString');

  const {
    className,
    highlightClassName,
    highlightStringProps,
    highlightStyle,
    highlightTerm,
    stringProps,
    style,
    text,
  } = props;
  const isNativeWind = isNativeWindInstalled();

  const defaultHighlightStyle = {
    backgroundColor: '#FFE4B5',
  };

  if (!highlightTerm.trim()) {
    return (
      <String
        className={isNativeWind ? className : undefined}
        style={isNativeWind ? undefined : style}
        {...stringProps}
      >
        {text}
      </String>
    );
  }

  const escapedHighlightTerm = highlightTerm.replace(/[$()*+.?[\\\]^{|}]/g, '\\$&');
  const regex = new RegExp(`(${escapedHighlightTerm})`, 'gi');
  const parts = text.split(regex);

  return (
    <String useFastText={false} {...classNameHandler(className)} {...styleHandler({ style })} {...stringProps}>
      {parts.map((part, i) => {
        if (part.toLowerCase() === highlightTerm.toLowerCase()) {
          return (
            <String
              key={i}
              useFastText={false}
              {...classNameHandler(cn({ 'bg-[#FFE4B5]': !highlightClassName }, highlightClassName))}
              {...styleHandler({ defaultStyle: highlightStyle ?? defaultHighlightStyle, style: highlightStyle })}
              {...highlightStringProps}
            >
              {part}
            </String>
          );
        }
        return part;
      })}
    </String>
  );
}

HighlightString.displayName = 'HighlightString';

export default memo(HighlightString);
