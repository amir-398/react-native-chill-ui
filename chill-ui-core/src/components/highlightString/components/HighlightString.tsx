import { memo } from 'react';
import { String } from '@components/string';
import { HighlightStringPropsTw } from '@types';
import { classNamePropsHandler, classNameHandler, styleHandler } from '@utils';

import { styles } from '../styles/HightlightString.styles';
import { twStyles } from '../styles/HightlightString.variants';

/**
 * The `<HighlightString />` component highlights specific terms within text.
 * Supports both NativeWind and traditional styling approaches.
 *
 *
 * <!-- STORYBOOK_IMPORT_START
 * ```tsx
 * import { HighlightString } from 'react-native-chill-ui';
 * ```
 * STORYBOOK_IMPORT_END -->
 *
 * @example
 * ```tsx
 * <HighlightString
 *   content="Hello world, welcome to the world of programming"
 *   highlightTerm="world"
 *   className="text-base text-gray-800"
 *   highlightClassName="bg-yellow-200 font-bold"
 * />
 * ```
 *
 * @param className - Custom CSS classes for the container (NativeWind)
 * @param content - The full text to display
 * @param highlightClassName - Custom CSS classes for highlighted text (NativeWind)
 * @param highlightStringProps - Props for the highlighted string component
 * @param highlightStyle - Custom styles for highlighted text
 * @param highlightTerm - The term to highlight within the text
 * @param stringProps - Props for the main string component
 * @param style - Custom styles for the container
 */
function HighlightString(props: HighlightStringPropsTw) {
  classNamePropsHandler(props, 'HighlightString');

  const {
    className,
    content,
    highlightClassName,
    highlightStringProps,
    highlightStyle,
    highlightTerm,
    stringProps,
    style,
  } = props;

  if (!highlightTerm.trim()) {
    return (
      <String {...classNameHandler(className)} {...styleHandler({ style })} {...stringProps}>
        {content}
      </String>
    );
  }

  const escapedHighlightTerm = highlightTerm.replace(/[$()*+.?[\\\]^{|}]/g, '\\$&');
  const regex = new RegExp(`(${escapedHighlightTerm})`, 'gi');
  const parts = content.split(regex);

  return (
    <String useFastText={false} {...classNameHandler(className)} {...styleHandler({ style })} {...stringProps}>
      {parts.map((part, i) => {
        if (part.toLowerCase() === highlightTerm.toLowerCase()) {
          return (
            <String
              key={i}
              useFastText={false}
              {...classNameHandler(highlightClassName ?? twStyles.defaultHighlightClassName)}
              {...styleHandler({ defaultStyle: highlightStyle ?? styles.defaultHighlightStyle, style: highlightStyle })}
              {...highlightStringProps}
            >
              {part}
            </String>
          );
        }
        return part;
      })}
    </String>
  );
}

HighlightString.displayName = 'HighlightString';

export default memo(HighlightString);
