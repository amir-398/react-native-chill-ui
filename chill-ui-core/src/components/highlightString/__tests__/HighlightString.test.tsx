import React from 'react';
import { render, screen } from '@testing-library/react-native';

import HighlightString from '../components/HighlightString';

// Mock the String components
jest.mock('../../string', () => ({
  String: ({ children, ...props }: any) => {
    const { Text } = require('react-native');
    return <Text {...props}>{children}</Text>;
  },
}));

// Mock the hybrid utils
jest.mock('@utils', () => ({
  classNameHandler: (className: string) => ({ className }),
  classNamePropsHandler: jest.fn(),
  styleHandler: ({ defaultStyle, style }: any) => ({ style: style || defaultStyle }),
}));

describe('HighlightString Component', () => {
  const defaultProps = {
    content: 'Hello world, welcome to the world of programming',
    highlightTerm: 'world',
  };

  describe('Basic Functionality', () => {
    it('renders without highlighting when highlightTerm is empty', () => {
      render(<HighlightString {...defaultProps} highlightTerm="" />);
      expect(screen.getByText(defaultProps.content)).toBeTruthy();
    });

    it('highlights multiple occurrences of term', () => {
      render(<HighlightString {...defaultProps} />);
      const highlightedTexts = screen.getAllByText('world');
      expect(highlightedTexts).toHaveLength(2);
    });

    it('handles case insensitive highlighting', () => {
      render(<HighlightString content="Hello WORLD" highlightTerm="world" />);
      expect(screen.getByText('WORLD')).toBeTruthy();
    });

    it('escapes special regex characters', () => {
      const content = 'Test $100 + 50% = $150';
      render(<HighlightString content={content} highlightTerm="$100" />);
      expect(screen.getByText('$100')).toBeTruthy();
    });
  });

  describe('Styling', () => {
    it('applies custom className and highlightClassName', () => {
      render(<HighlightString {...defaultProps} className="custom-class" highlightClassName="highlight-class" />);
      const textElement = screen.getByText(defaultProps.content);
      const highlightedElements = screen.getAllByText('world');

      expect(textElement.props.className).toContain('custom-class');
      highlightedElements.forEach(element => {
        expect(element.props.className).toContain('highlight-class');
      });
    });

    it('applies custom style and highlightStyle', () => {
      const customStyle = { color: 'red', fontSize: 18 };
      const customHighlightStyle = { backgroundColor: 'yellow', fontWeight: 'bold' };

      render(<HighlightString {...defaultProps} style={customStyle} highlightStyle={customHighlightStyle} />);

      const textElement = screen.getByText(defaultProps.content);
      const highlightedElements = screen.getAllByText('world');

      expect(textElement.props.style).toEqual(expect.objectContaining(customStyle));
      highlightedElements.forEach(element => {
        expect(element.props.style).toEqual(expect.objectContaining(customHighlightStyle));
      });
    });
  });

  describe('Edge Cases', () => {
    it('handles empty content', () => {
      render(<HighlightString content="" highlightTerm="test" />);
      expect(screen.getByText('')).toBeTruthy();
    });

    it('handles unicode characters', () => {
      const content = 'Hello 世界, welcome to the 世界 of programming';
      render(<HighlightString content={content} highlightTerm="世界" />);
      const highlightedTexts = screen.getAllByText('世界');
      expect(highlightedTexts).toHaveLength(2);
    });
  });

  describe('Snapshots', () => {
    it('matches snapshot for basic highlighting', () => {
      const tree = render(<HighlightString {...defaultProps} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('matches snapshot with custom styling', () => {
      const tree = render(
        <HighlightString
          {...defaultProps}
          className="text-lg text-blue-500"
          highlightClassName="bg-yellow-200 font-bold"
          style={{ fontSize: 16 }}
          highlightStyle={{ backgroundColor: '#FFE4B5' }}
        />,
      );
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('matches snapshot with no highlighting', () => {
      const tree = render(<HighlightString {...defaultProps} highlightTerm="" />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('matches snapshot with special characters', () => {
      const tree = render(<HighlightString content="Test $100 + 50% = $150" highlightTerm="$100" />);
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe('Integration', () => {
    it('works with stringProps and highlightStringProps', () => {
      const stringProps = { colorVariant: 'primary' as const, size: 'lg' as const };
      const highlightStringProps = { colorVariant: 'secondary' as const, size: 'sm' as const };

      render(
        <HighlightString {...defaultProps} stringProps={stringProps} highlightStringProps={highlightStringProps} />,
      );

      const textElement = screen.getByText(defaultProps.content);
      const highlightedElements = screen.getAllByText('world');

      expect(textElement.props).toEqual(expect.objectContaining(stringProps));
      highlightedElements.forEach(element => {
        expect(element.props).toEqual(expect.objectContaining(highlightStringProps));
      });
    });

    it('handles complex content with multiple highlights', () => {
      const content = 'The quick brown fox jumps over the lazy dog. The fox is quick.';
      render(<HighlightString content={content} highlightTerm="fox" />);

      const highlightedElements = screen.getAllByText('fox');
      expect(highlightedElements).toHaveLength(2);
      expect(screen.getByText(content)).toBeTruthy();
    });

    it('maintains performance with large content', () => {
      const largeContent = `${'word '.repeat(100)}target ${'word '.repeat(100)}`;
      const startTime = Date.now();

      render(<HighlightString content={largeContent} highlightTerm="target" />);

      const endTime = Date.now();
      expect(endTime - startTime).toBeLessThan(100); // Should render in less than 100ms
      expect(screen.getByText('target')).toBeTruthy();
    });
  });
});
