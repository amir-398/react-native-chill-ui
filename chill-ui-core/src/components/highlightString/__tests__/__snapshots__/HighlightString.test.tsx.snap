// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightString Component Snapshots matches snapshot for basic highlighting 1`] = `
<Text
  useFastText={false}
>
  Hello 
  <Text
    className="bg-[#FFE4B5]"
    style={
      {
        "backgroundColor": "#FFE4B5",
      }
    }
    useFastText={false}
  >
    world
  </Text>
  , welcome to the 
  <Text
    className="bg-[#FFE4B5]"
    style={
      {
        "backgroundColor": "#FFE4B5",
      }
    }
    useFastText={false}
  >
    world
  </Text>
   of programming
</Text>
`;

exports[`HighlightString Component Snapshots matches snapshot with custom styling 1`] = `
<Text
  className="text-lg text-blue-500"
  style={
    {
      "fontSize": 16,
    }
  }
  useFastText={false}
>
  Hello 
  <Text
    className="bg-yellow-200 font-bold"
    style={
      {
        "backgroundColor": "#FFE4B5",
      }
    }
    useFastText={false}
  >
    world
  </Text>
  , welcome to the 
  <Text
    className="bg-yellow-200 font-bold"
    style={
      {
        "backgroundColor": "#FFE4B5",
      }
    }
    useFastText={false}
  >
    world
  </Text>
   of programming
</Text>
`;

exports[`HighlightString Component Snapshots matches snapshot with no highlighting 1`] = `
<Text>
  Hello world, welcome to the world of programming
</Text>
`;

exports[`HighlightString Component Snapshots matches snapshot with special characters 1`] = `
<Text
  useFastText={false}
>
  Test 
  <Text
    className="bg-[#FFE4B5]"
    style={
      {
        "backgroundColor": "#FFE4B5",
      }
    }
    useFastText={false}
  >
    $100
  </Text>
   + 50% = $150
</Text>
`;
